File: DriftGame/uracer-desktop/src/com/bitfire/uracer/utils/UIUtils.java
package com.bitfire.uracer.utils;

import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.CheckBox;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.List;
import com.badlogic.gdx.scenes.scene2d.ui.ScrollPane;
import com.badlogic.gdx.scenes.scene2d.ui.SelectBox;
import com.badlogic.gdx.scenes.scene2d.ui.Slider;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.viewport.FitViewport;
import com.bitfire.uracer.URacer;
import com.bitfire.uracer.resources.Art;

public final class UIUtils {

    public static final boolean debug = false;

    private UIUtils() {
    }

    public static Stage newFittedStage() {
        return new Stage(new FitViewport(ScaleUtils.PlayWidth, ScaleUtils.PlayHeight));
    }

    public static TextButton newTextButton(String text, ClickListener listener) {
        TextButton btn = new TextButton(text, Art.scrSkin);
        btn.addListener(listener);
        return btn;
    }

    public static CheckBox newCheckBox(String text, boolean checked, ChangeListener listener) {
        CheckBox cb = new CheckBox(" " + text, Art.scrSkin);
        cb.setChecked(checked);
        if (listener != null) {
            cb.addListener(listener);
        }
        return cb;
    }

    public static SelectBox<String> newSelectBox(String[] items, ChangeListener listener) {
        SelectBox<String> sb = new SelectBox<>(Art.scrSkin);
        if (listener != null) {
            sb.addListener(listener);
        }

        sb.setItems(items);
        return sb;
    }

    public static List<String> newListBox(String[] items) {
        return newListBox(items, null);
    }

    public static List<String> newListBox(String[] items, ChangeListener listener) {
        List<String> list = new List<>(Art.scrSkin);
        if (listener != null) {
            list.addListener(listener);
        }

        list.setItems(items);
        return list;
    }

    public static Slider newSlider(float min, float max, float step, float value, ChangeListener listener) {
        Slider s = new Slider(min, max, step, false, Art.scrSkin);
        s.setValue(value);
        if (listener != null) {
            s.addListener(listener);
        }
        return s;
    }

    public static Label newLabel(String text, boolean wrap) {
        Label l = new Label(text, Art.scrSkin);
        l.setWrap(wrap);
        return l;
    }

    public static ScrollPane newScrollPane() {
        return new ScrollPane(null, Art.scrSkin);
    }

    public static TextButton newButton(String text, ClickListener listener) {
        TextButton b = new TextButton(text, Art.scrSkin);
        if (listener != null) {
            b.addListener(listener);
        }
        return b;
    }

    public static Table newTable() {
        Table t = new Table();
        if (debug) {
            t.debug();
        }
        return t;
    }

    public static Window newWindow(String title) {
        Window w = new Window(title, Art.scrSkin);
        if (debug) {
            w.debug();
        }
        return w;
    }

    public static Label newVersionInfoLabel() {
        return UIUtils.newLabel("uRacer " + URacer.versionInfo, false);
    }
}

File: DriftGame/uracer-desktop/src/com/bitfire/uracer/utils/UIUtils.kt
package com.bitfire.uracer.utils

import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.scenes.scene2d.ui.List
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.utils.viewport.FitViewport
import com.bitfire.uracer.URacer
import com.bitfire.uracer.resources.Art

object UIUtils {

    const val DEBUG = false

    @JvmStatic
    fun newFittedStage() = Stage(FitViewport(ScaleUtils.PlayWidth.toFloat(), ScaleUtils.PlayHeight.toFloat()))

    @JvmStatic
    fun newTextButton(text: String, listener: ClickListener): TextButton {
        val btn = TextButton(text, Art.scrSkin)
        btn.addListener(listener)
        return btn
    }

    @JvmStatic
    fun newCheckBox(text: String, checked: Boolean, listener: ChangeListener?): CheckBox {
        val cb = CheckBox(" $text", Art.scrSkin)
        cb.setChecked(checked)
        if (listener != null) {
            cb.addListener(listener)
        }
        return cb
    }

    @JvmStatic
    fun newSelectBox(items: Array<String>, listener: ChangeListener?): SelectBox<String> {
        val sb = SelectBox<String>(Art.scrSkin)
        if (listener != null) {
            sb.addListener(listener)
        }

        sb.setItems(*items)
        return sb
    }

    @JvmStatic
    @JvmOverloads
    fun newListBox(items: Array<String>, listener: ChangeListener? = null): List<String> {
        val list = List<String>(Art.scrSkin)
        if (listener != null) {
            list.addListener(listener)
        }

        list.setItems(*items)
        return list
    }

    @JvmStatic
    fun newSlider(min: Float, max: Float, step: Float, value: Float, listener: ChangeListener?): Slider {
        val s = Slider(min, max, step, false, Art.scrSkin)
        s.value = value
        if (listener != null) {
            s.addListener(listener)
        }
        return s
    }

    @JvmStatic
    fun newLabel(text: String, wrap: Boolean): Label {
        val l = Label(text, Art.scrSkin)
        l.setWrap(wrap)
        return l
    }

    @JvmStatic
    fun newScrollPane() = ScrollPane(null, Art.scrSkin)

    @JvmStatic
    fun newButton(text: String, listener: ClickListener?): TextButton {
        val b = TextButton(text, Art.scrSkin)
        if (listener != null) {
            b.addListener(listener)
        }
        return b
    }

    @JvmStatic
    fun newTable(): Table {
        val t = Table()
        if (DEBUG) t.debug()
        return t
    }

    @JvmStatic
    fun newWindow(title: String): Window {
        val w = Window(title, Art.scrSkin)
        if (DEBUG) w.debug()
        return w
    }

    @JvmStatic
    fun newVersionInfoLabel() = newLabel("uRacer ${URacer.versionInfo}", false)

}
